### PR TITLE
Update router state properties directly

### PR DESCRIPTION
**Pull request by timdorr**
This will address the last checkbox on #3445 

It will also shut me up about this comment https://github.com/reactjs/react-router/issues/3445#issuecomment-218000037 😄

### CODE REVIEW

File: modules/Router.js [hunk 1]
```diff
@@ -91,7 +90,7 @@ const Router = React.createClass({
       } else {
         // Keep the identity of this.router because of a caveat in ContextUtils:
         // they only work if the object identity is preserved.
-        assign(this.router, this.createRouterObject(state))
+        updateRouterObjectState(this.router, state)
```
Comment (taion) [inline modules/Router.js]: Maybe just inline this?

Comment (taion) [inline modules/Router.js]: Or call it something like `assignRouterState` and use it in `createRouterObject` as well.

Comment (timdorr) [inline modules/Router.js]: I wanted to keep the shape of `router` in one place, rather than having coupling to it's shape spread into two places.

Comment (timdorr) [inline modules/Router.js]: > use it in createRouterObject as well

Oh, that's a good point. One sec...

**Comment by taion** [comment]
LGTM

**Pull request merged**